### DISCUSSION TITLE
ImportError: cannot import name 'pkcs7'

### DISCUSSION THREAD

**Discussion by CLSMCSMII** [polhenarejos/acme_email#2] [Q&A] [answered]
After git clone and run setup.py successfully, I tried this command

python3 cli.py cert --config-dir . --work-dir . --logs-dir . -e clxxxxx@icloud.com --contact panxxx@xxxxxx --usage digitalSignature --usage keyEncipherment
(both emails are valid)

Then I got this error:

An unexpected error occurred:
ImportError: cannot import name 'pkcs7'
Ask for help or search for solutions at https://community.letsencrypt.org. See the logfile /tmp/tmphkn09j3p/log or re-run Certbot with -v for more details.

And.. tail /tmp/tmphkn09j3p/log

  File "/home/tempuser/acme/acme_email/certbot_castle/plugins/imap.py", line 22, in <module>
    from cryptography.hazmat.primitives.serialization import pkcs7
ImportError: cannot import name 'pkcs7'
2021-07-16 10:55:46,200:ERROR:certbot._internal.log:An unexpected error occurred:
2021-07-16 10:55:46,200:ERROR:certbot._internal.log:ImportError: cannot import name 'pkcs7'

Already install pkcs7 with this command "pip3 install pkcs7" still got the same error.

Any idea?

**Comment by DDvO** [comment]
I just found that I'm not the first one running into this issue - see also #14.
Shame that this issue has not been answered yet.

I  know very little about Python, but looks like the issue is due to a version incompatibility with https://cryptography.io/en/latest/hazmat/primitives/asymmetric/serialization/

BTW, that page contains a nice warning:
> This is a “Hazardous Materials” module. You should ONLY use it if you’re 100% absolutely sure that you know what you’re doing because this module is full of land mines, dragons, and dinosaurs with laser guns.

**Comment by DDvO** [comment, marked as answer]
@polhenarejos,  for some reason `setup.py` script does not get the version of the `cryptography` module right.
I got:
```
Searching for cryptography>=2.1.4
Reading https://pypi.org/simple/cryptography/
Downloading https://files.pythonhosted.org/packages/10/91/90b8d4cd611ac2aa526290ae4b4285aa5ea57ee191c63c2f3d04170d7683/cryptography-35.0.0.tar.gz#sha256=9933f28f70d0517686bd7de36166dda42094eac49415459d9bdf5e7df3e0086d
Best match: cryptography 35.0.0
Processing cryptography-35.0.0.tar.gz
```
since on my Debian 10 system, version 2.8 was installed, and this does not include pkcs7 support.
According to https://github.com/pyca/cryptography/blob/main/CHANGELOG.rst it was added in version 3.0 and extended in 3.1.

After manually installing `python3-cryptography (3.3.2-1)` from Debian 11, the issue went away.

**Comment by polhenarejos** [comment]
I was not aware of this discussion. It is preferable to open an issue for bugs/errors.

It is really strange your pip installed a 4 years old version. Probably it was a system installation done some years ago.
Did you use a specific venv for acme_email?

BTW, I uploaded the `setup.py` to add `cryptography`. A `pip3 install -U .` should do the trick.
